File: AoC_2022/AoC_2022_01/main.cpp
////////////////////////////////////////////////////
//              Advent of Code 2022               //
//            Day 1: Calorie Counting             //
//            Solution by Lorin Lange             //
////////////////////////////////////////////////////

#include <iostream>
#include <fstream>
#include <string>
#include <algorithm>
#include <vector>

#define FILE_NAME "test_input.txt"

using namespace std;

int main() {
    ifstream file;
    file.open(FILE_NAME);
    vector<int> list;
    if(file.is_open()) {
        int sum = 0;
        string line;
        while(file) {
            getline(file, line);
            if(line.empty()) {
                list.push_back(sum);
                sum = 0;
            } else {
                sum += stoi(line);
            }
        }
    }

    sort(list.begin(), list.end(), greater<int>());

    cout << "Part 1: " << list[0] << endl;
    cout << "Part 2: " << list[0] + list[1] + list[2] << endl;

    return EXIT_SUCCESS;
}

File: AoC_2022/AoC_2022_01/Main.hs
----------------------------------------------------
--              Advent of Code 2022               --
--            Day 1: Calorie Counting             --
--            Solution by Lorin Lange             --
----------------------------------------------------

module Main where

import Text.Read ( readMaybe )
import Data.List ( sortBy )
import Data.List.Split ( endBy )
import Data.Maybe ( fromMaybe )
import Data.Ord ( Down(Down), comparing )

getSums :: [Maybe Integer] -> [Integer]
getSums = sortBy (comparing Down) . map (sum . map (fromMaybe 0)) . endBy [Nothing]

main :: IO()
main = do sums <- getSums . map readMaybe . lines <$> readFile "input.txt"
          putStrLn $ "Part 1: " ++ show (head sums)
          putStrLn $ "Part 2: " ++ show (sum $ take 3 sums)
